src/main/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2.scala

```diff
@@ -0,0 +1,194 @@
+      relation: LogicalRelation,
+      joinCols: Set[AttributeReference],
+      allReferencedCols: Seq[Attribute]): LogicalPlan = {
+    val indexManager = Hyperspace
+      .getContext(spark)
+      .indexCollectionManager
+
+    // TODO: the following check only considers indexes in ACTIVE state for usage. Update
+    //  the code to support indexes in transitioning states as well.
+    //  See https://github.com/microsoft/hyperspace/issues/65.
+    val allIndexes = indexManager.getIndexes(Seq(Constants.States.ACTIVE))
+
+    val allCols = (joinCols ++ allReferencedCols.toSet).map(_.asInstanceOf[AttributeReference])
+
+    val allReqdCols =
+      relation.outputSet
+        .filter(c => contains(allCols, c.asInstanceOf[AttributeReference]))
+        .map(_.name)
+    val reqdIdxCols = relation.outputSet
+      .filter(
+        c =>
+          contains(
+            joinCols.map((col: Attribute) => col.asInstanceOf[AttributeReference]),
+            c.asInstanceOf[AttributeReference]))
+      .map(_.name)
+
+    val usableIndexes = allIndexes
+      .filter { index =>
+        index.config.indexedColumns.toSet.equals(reqdIdxCols.toSet) &&
+        allReqdCols.forall(
+          (index.config.indexedColumns ++ index.config.includedColumns).contains(_))
+      }
+
+    usableIndexes match {
+      case Nil => relation
+      case _ =>
+        RuleUtils.getCandidateIndexes(spark, usableIndexes, relation) match {
+          case Nil => relation
+          case candidateIndex :: _ =>
+            RuleUtils.transformPlanToUseIndex(spark, candidateIndex, relation, true)
+        }
+    }
+  }
+
+  private def updatePlan(join: Join): Join = {
+    val newLeft: LogicalPlan = updateIfSupported(join.left, join.condition.get)
+    val newRight: LogicalPlan = updateIfSupported(join.right, join.condition.get)
+    join.copy(left = newLeft, right = newRight)
+  }
+
+  private def updateIfSupported(plan: LogicalPlan, condition: Expression): LogicalPlan = {
+    // Either left or right or both sides should contain half of references.
+    // All references for any side, either left or right, should all come from the same leaf node.
+    //
+    // E.g. if join condition refers to columns A, B, C, D, two columns should come from left, 2
+    // from right. This would be by default true. Nothing to do in this part. Let's say A,B
+    // come from left. C,D come from right. The requirement is both A and B should come from the
+    // same leaf node on left. Same for C and D. Both should come from same leaf node on right.
+
+    val joinCols = condition.references.map(_.asInstanceOf[AttributeReference]).toSet
+    val eligibleBaseRelations = plan.collectLeaves().filter {
+      case relation: LogicalRelation => relation.output.exists(contains(joinCols, _))
+      case _ => false
+    }
+
+    eligibleBaseRelations match {
+      case Seq(r: LogicalRelation)
+          if r.output.toSet.count(c => contains(joinCols, c)) * 2 == joinCols.size =>
```

Comment:
thanks @imback82 , in this case we should not use the index on any side. It will not eliminate shuffle because left side and right side have different partitioning requirements.

Reply:
but we should be able to apply, no?

Reply:
I don't think it will be able to apply: Here's the example

```scala
  spark.conf.set("spark.sql.autoBroadcastJoinThreshold", "-1")
  import spark.implicits._
  spark.sql("DROP TABLE if exists t1").show
  spark.sql("DROP TABLE if exists t2").show
  val df1 = (0 until 100).map(i => (i % 5, i % 13, i.toString)).toDF("i1", "j1", "k1")
  val df2 = (0 until 100).map(i => (i % 5, i % 13, i.toString)).toDF("i2", "j2", "k2")
  df1.write.format("parquet").bucketBy(4, "i1").saveAsTable("t1")                            // Bucket by i1
  df2.write.format("parquet").bucketBy(4, "i2", "j2").saveAsTable("t2")                    // Bucket by i2, j2
  val t1 = spark.table("t1")
  val t2 = spark.table("t2")
  val joined = t1.join(t2, t1("i1") === t2("i2") && t1("i1") === t2("j2"))                 // i1 = i2 and i1 = j2
  joined.explain(true)
  spark.sql("drop table t1")
  spark.sql("drop table t2")
```
Result:
```
== Optimized Logical Plan ==
Join Inner, ((i1#34 = i2#40) && (i1#34 = j2#41))
:- Filter isnotnull(i1#34)
:  +- Relation[i1#34,j1#35,k1#36] parquet
+- Filter (((i2#40 = j2#41) && isnotnull(i2#40)) && isnotnull(j2#41))
   +- Relation[i2#40,j2#41,k2#42] parquet

== Physical Plan ==
*(5) SortMergeJoin [i1#34, i1#34], [i2#40, j2#41], Inner
:- *(2) Sort [i1#34 ASC NULLS FIRST, i1#34 ASC NULLS FIRST], false, 0
:  +- Exchange hashpartitioning(i1#34, i1#34, 200)
:     +- *(1) Project [i1#34, j1#35, k1#36]
:        +- *(1) Filter isnotnull(i1#34)
:           +- *(1) FileScan parquet default.t1[i1#34,j1#35,k1#36] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/apdave/github/hyperspace-1/spark-warehouse/t1], PartitionFilters: [], PushedFilters: [IsNotNull(i1)], ReadSchema: struct<i1:int,j1:int,k1:string>, SelectedBucketsCount: 4 out of 4
+- *(4) Sort [i2#40 ASC NULLS FIRST, j2#41 ASC NULLS FIRST], false, 0
   +- Exchange hashpartitioning(i2#40, j2#41, 200)
      +- *(3) Project [i2#40, j2#41, k2#42]
         +- *(3) Filter (((i2#40 = j2#41) && isnotnull(i2#40)) && isnotnull(j2#41))
            +- *(3) FileScan parquet default.t2[i2#40,j2#41,k2#42] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/apdave/github/hyperspace-1/spark-warehouse/t2], PartitionFilters: [], PushedFilters: [IsNotNull(i2), IsNotNull(j2)], ReadSchema: struct<i2:int,j2:int,k2:string>, SelectedBucketsCount: 4 out of 4
```
Neither side is able to utilize bucketing.

Reply:
Thanks @apoorvedave1 for looking into this! What if you set `spark.sql.shuffle.partitions` to same as your bucket number?

Reply:
oh ok so this worked for one side, thanks @imback82 :
```scala
*(4) SortMergeJoin [i1#40, i1#40], [i2#46, j2#47], Inner
:- *(2) Sort [i1#40 ASC NULLS FIRST, i1#40 ASC NULLS FIRST], false, 0
:  +- Exchange hashpartitioning(i1#40, i1#40, 4)
:     +- *(1) Project [i1#40, j1#41, k1#42]
:        +- *(1) Filter isnotnull(i1#40)
:           +- *(1) FileScan parquet default.t1[i1#40,j1#41,k1#42] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/apdave/spark-warehouse/t1], PartitionFilters: [], PushedFilters: [IsNotNull(i1)], ReadSchema: struct<i1:int,j1:int,k1:string>, SelectedBucketsCount: 4 out of 4
+- *(3) Sort [i2#46 ASC NULLS FIRST, j2#47 ASC NULLS FIRST], false, 0
   +- *(3) Project [i2#46, j2#47, k2#48]
      +- *(3) Filter (((i2#46 = j2#47) && isnotnull(i2#46)) && isnotnull(j2#47))
         +- *(3) FileScan parquet default.t2[i2#46,j2#47,k2#48] Batched: true, Format: Parquet, Location: InMemoryFileIndex[file:/C:/Users/apdave/spark-warehouse/t2], PartitionFilters: [], PushedFilters: [IsNotNull(i2), IsNotNull(j2)], ReadSchema: struct<i2:int,j2:int,k2:string>, SelectedBucketsCount: 4 out of 4
```

So when the partitioning of both sides is different, then it becomes a requirement the numBuckets of index === `spark.sql.shuffle.partitions`. If partitioning of both sides is same, spark automatically removes shuffle without this constraint.

So now how should we handle this, do we pick index assuming the above requirement is satisfied? (I don't think we can enforce this as we only work on one side of the plan at a time). I guess we can use the index with the assumption, i don't think there is a downside to this.

Please let me know your thoughts. I have pushed a commit which supports picking index in this scenario for table 2.

Reply:
https://github.com/microsoft/hyperspace/pull/124/commits/70b45ce5186dbf89e9b0bde5e66d837987d13988

Reply:
So, it sounds like we should always check `spark.sql.shuffle.partitions` since we are updating one side independently?

Also, it seems like we need to check the other side of join if it's a bucketed table (not our index, but the user is reading from a bucketed table)? (If the number of bucket doesn't line up, it will create a shuffle on the either side).

One option to release is to have the current implementation, and list the current limitation which you can address as a follow up.

Reply:
> So, it sounds like we should always check `spark.sql.shuffle.partitions` since we are updating one side independently?

but if both sides have equal numBuckets (`t1.numBuckets == t2.numBuckets`), and `numBuckets != spark.sql.shuffle.partitions`, spark can still eliminate shuffle. In such a scenario if we add the above check you mentioned, we may not pick the index and lose potential improvement, right?

I can capture all this in an issue and add the link here. Thanks.

Reply:
> Also, it seems like we need to check the other side of join if it's a bucketed table (not our index, but the user is reading from a bucketed table)? (If the number of bucket doesn't line up, it will create a shuffle on the either side).

It seems like we might need to add this check to JoinIndexRanker as well. The ranker prefers indexes with equal buckets but if equal buckets are not found, it will pick indexes with unequal buckets. One side will eliminate shuffle, other will not.

Reply:
https://github.com/microsoft/hyperspace/issues/237

Reply:
> but if both sides have equal numBuckets (`t1.numBuckets == t2.numBuckets`), and `numBuckets != spark.sql.shuffle.partitions`, spark can still eliminate shuffle. In such a scenario if we add the above check you mentioned, we may not pick the index and lose potential improvement, right?

The point is that we should be more careful since we are updating one side independently in join v2. With the current implementation, it appears that we are not sure what's going to happen.

Reply:
oh fair enough. I can start another PR for targetting #237 with more checks:
- if #buckets and partitioning is same for t1 and t2
- if #buckets is different and one side matches with `spark.sql.shuffle.partitions`